aws_signing_helper/pkcs11_signer.go

```diff
@@ -187,10 +440,18 @@ func (pkcs11Signer *PKCS11Signer) Sign(rand io.Reader, digest []byte, opts crypt
 	module := pkcs11Signer.module
 	session := pkcs11Signer.session
 	privateKeyHandle := pkcs11Signer.privateKeyHandle
-	cert := pkcs11Signer.cert
-
+	keyType := pkcs11Signer.keyType
+
+	// XXX: If you use this outside the context of IAM RA, be aware that
+	// you'll want to use something other than SHA256 in many cases.
+	// For TLSv1.3 the hash needs to precisely match the bit size of the
+	// curve, IIRC. And you'll need RSA-PSS too. You might find that
+	// ThalesIgnite/crypto11 has some of that.
+	// e.g. https://github.com/ThalesIgnite/crypto11/blob/master/rsa.go#L230
 	var mechanism uint
-	if cert.PublicKeyAlgorithm == x509.ECDSA {
+	if keyType == pkcs11.CKK_EC {
+		hash := sha256.Sum256(digest)
+		digest = hash[:]
 		mechanism = pkcs11.CKM_ECDSA
```

Comment:
Minor, but would things be more consistent between EC and RSA keys if we used `CKM_ECDSA_{HASH_FUNCTION}` here instead?

Reply:
SoftHSM doesn't seem to support that:
```
C_SignInit
  IN: hSession = S1
  IN: pMechanism = {
	mechanism: CKM_0x00001044
	pParameter: (0) NULL
      }
  IN: hKey = H2
C_SignInit = CKR_MECHANISM_INVALID
2023/06/21 16:45:42 unable to sign the digest

```

Reply:
I do seem to have only supported SHA256 though, while the RSA case explicitly supports others. I'll look at fixing that.

Reply:
Fixed in commit 65e3816